opentimelineview/timeline_widget.py

```diff
@@ -344,24 +353,262 @@ def counteract_zoom(self, zoom_level=1.0):
 
 
 class TimeSlider(QtWidgets.QGraphicsRectItem):
+
     def __init__(self, *args, **kwargs):
         super(TimeSlider, self).__init__(*args, **kwargs)
         self.setBrush(QtGui.QBrush(QtGui.QColor(64, 78, 87, 255)))
         pen = QtGui.QPen()
         pen.setWidth(0)
         self.setPen(pen)
+        self._ruler = None
+
+    def mousePressEvent(self, mouse_event):
+        pos = self.mapToScene(mouse_event.pos())
+        self._ruler.setPos(QtCore.QPointF(pos.x(),
+            TIME_SLIDER_HEIGHT - MARKER_SIZE))
+        self._ruler.update_frame()
+
+        super(TimeSlider, self).mousePressEvent(mouse_event)     
+
+    def add_ruler(self, ruler):
+        self._ruler = ruler        
+
+
+class FrameNumber(QtWidgets.QGraphicsRectItem):
+
+    def __init__(self, text, position, *args, **kwargs):
+        super(FrameNumber, self).__init__(*args, **kwargs)
+        self.frameNumber = QtWidgets.QGraphicsSimpleTextItem(self)
+        self.frameNumber.setText("%s" % text)
+        self.setBrush(
+            QtGui.QBrush(QtGui.QColor(5, 55, 0, 255))
+        )
+        self.setPen(
+            QtCore.Qt.NoPen
+        )
+        self.frameNumber.setBrush(
+            QtGui.QBrush(QtGui.QColor(25, 255, 10, 255))
+        )
+        # if position < 0 then the frameNumber will appear on the left side
+        # of the ruler
+        self.position = position
+
+    def setText(self, txt, highlight =  False):
+        if txt:
+            self.show()
+            self.frameNumber.setText("%s" % txt)
+            rect = self.frameNumber.boundingRect()
+            self.setRect(self.frameNumber.boundingRect())
+            if highlight:
+                # paint with a different color when on 
+                # the first frame of a clip
+                self.setBrush(
+                    QtGui.QBrush(QtGui.QColor(55, 5, 0, 120))
+                )
+                self.frameNumber.setBrush(
+                    QtGui.QBrush(QtGui.QColor(255, 250, 250, 255))
+                )
+            else:
+                self.frameNumber.setBrush(
+                    QtGui.QBrush(QtGui.QColor(25, 255, 10, 255))
+                )
+                self.setBrush(
+                    QtGui.QBrush(QtGui.QColor(5, 55, 0, 120))
+                )
+            if self.position <0 :
+                self.setX(-rect.width()-2)
+            else :
+                self.setX(2)
+        else:
+            self.hide()
+
+
+class Ruler(QtWidgets.QGraphicsPolygonItem):
+
+    time_space = OrderedDict ([ # @TODO pending on Global Space implementation
+                                #("external_space", "External Space"),
+                                ("media_space", "Media Space"),
```

Comment:
@jminor what do you think these labels should be?

Reply:
I talked to josh - we concluded following the naming convention in the time-scopes document makes sense for now at  least.  That would be "Media Space" and "Trimmed Space".  We can adapt them later if we decide to change the names.